docs/02-app/01-building-your-application/09-authentication/index.mdx

```diff
@@ -161,730 +570,1143 @@ export default async function handler(req, res) {
 
-import { authenticate } from '@/app/lib/actions'
-import { useFormState, useFormStatus } from 'react-dom'
+```ts filename="app/actions/auth.ts" switcher
+import { createSession } from '@/app/lib/session'
 
-export default function Page() {
-  const [errorMessage, dispatch] = useFormState(authenticate, undefined)
+export async function signup(state: FormState, formData: FormData) {
+  // Previous steps:
+  // 1. Validate form fields
+  // 2. Prepare data for insertion into database
+  // 3. Insert the user into the database or call an Library API
 
-  return (
-    <form action={dispatch}>
-      <input type="email" name="email" placeholder="Email" required />
-      <input type="password" name="password" placeholder="Password" required />
-      <div>{errorMessage && <p>{errorMessage}</p>}</div>
-      <LoginButton />
-    </form>
-  )
+  // Current steps:
+  // 4. Create user session
+  await createSession(user.id)
+  // 5. Redirect user
+  redirect('/profile')
 }
+```
 
-function LoginButton() {
-  const { pending } = useFormStatus()
+```js filename="app/actions/auth.js" switcher
+import { createSession } from '@/app/lib/session'
 
-  const handleClick = (event) => {
-    if (pending) {
-      event.preventDefault()
-    }
-  }
+export async function signup(state, formData) {
+  // Previous steps:
+  // 1. Validate form fields
+  // 2. Prepare data for insertion into database
+  // 3. Insert the user into the database or call an Library API
 
-  return (
-    <button aria-disabled={pending} type="submit" onClick={handleClick}>
-      Login
-    </button>
-  )
+  // Current steps:
+  // 4. Create user session
+  await createSession(user.id)
+  // 5. Redirect user
+  redirect('/profile')
 }
 ```
 
-</AppOnly>
+> **Tips**:
+>
+> - **Cookies should be set on the server** to prevent client-side tampering.
+> - 🎥 Watch: Learn more about stateless sessions and authentication with Next.js → [YouTube (11 minutes)](https://www.youtube.com/watch?v=DJvM2lSPn6w).
 
-For a more streamlined authentication setup in Next.js projects, especially when offering multiple login methods, consider using a comprehensive [authentication solution](#examples).
+#### Updating (or refreshing) sessions
 
-## Authorization
+You can also extend the session's expiration time. This is useful for keeping the user logged in after they access the application again. For example:
 
-Once a user is authenticated, you'll need to ensure the user is allowed to visit certain routes, and perform operations such as mutating data with Server Actions and calling Route Handlers.
+```ts filename="app/actions/session.ts" switcher
+import 'server-only'
+import { cookies } from 'next/headers'
 
-### Protecting Routes with Middleware
```

Comment:
@delbaoliveira Hi, why was the "Protecting Routes with Middleware" section removed?

Reply:
Hey, it's now called **Optimistic checks with Middleware*.